File: scripts/test_mcp.py
#!/usr/bin/env python3
"""
Test script for Universal MCP Implementation
Tests MCP client and server with multiple AI agents
"""

import sys
import json
import subprocess
from pathlib import Path

# ANSI colors
class Colors:
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'

def print_banner():
    print(f"\n{Colors.BOLD}{'='*60}{Colors.ENDC}")
    print(f"{Colors.BOLD}ELYX Universal MCP Test Suite{Colors.ENDC}")
    print(f"{Colors.BOLD}{'='*60}{Colors.ENDC}\n")

def test_step(name, status="pending"):
    """Print test step status"""
    icons = {"pending": "⏳", "running": "🔄", "passed": "✅", "failed": "❌"}
    status_str = f"{icons.get(status, '•')} {name}"
    
    if status == "passed":
        print(f"{Colors.OKGREEN}{status_str}{Colors.ENDC}")
    elif status == "failed":
        print(f"{Colors.FAIL}{status_str}{Colors.ENDC}")
    else:
        print(status_str)
    
    return status == "passed"

def test_mcp_server_direct():
    """Test MCP server via direct stdin"""
    test_step("Testing Email MCP Server (direct stdin)", "running")
    
    try:
        request = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "tools.list",
            "params": {}
        }
        
        server_path = Path("src/mcp-servers/email-mcp/index.py")
        if not server_path.exists():
            test_step("Email MCP Server not found", "failed")
            return False
        
        result = subprocess.run(
            ["python", str(server_path)],
            input=json.dumps(request),
            capture_output=True,
            text=True,
            timeout=10
        )
        
        if result.returncode != 0:
            test_step(f"Server error: {result.stderr[:200]}", "failed")
            return False
        
        response = json.loads(result.stdout)
        
        if "result" in response:
            tools = response["result"].get("tools", [])
            test_step(f"Email MCP Server responding ({len(tools)} methods)", "passed")
            return True
        elif "error" in response:
            test_step(f"Server error: {response['error']['message']}", "failed")
            return False
        else:
            test_step("Invalid response format", "failed")
            return False
            
    except subprocess.TimeoutExpired:
        test_step("Server timeout", "failed")
        return False
    except Exception as e:
        test_step(f"Error: {e}", "failed")
        return False

def test_mcp_client_import():
    """Test MCP client library import"""
    test_step("Testing MCP Client Library import", "running")
    
    try:
        from src.mcp_client import MCPClient, send_email
        test_step("MCP Client Library imported successfully", "passed")
        return True
    except ImportError as e:
        test_step(f"Import error: {e}", "failed")
        return False
    except Exception as e:
        test_step(f"Error: {e}", "failed")
        return False

def test_mcp_client_instantiation():
    """Test MCP client instantiation"""
    test_step("Testing MCP Client instantiation", "running")
    
    try:
        from src.mcp_client import MCPClient
        
        client = MCPClient("email", transport="stdio")
        
        if client.server == "email" and client.transport == "stdio":
            test_step("MCP Client instantiated correctly", "passed")
            return True
        else:
            test_step("Client configuration incorrect", "failed")
            return False
            
    except Exception as e:
        test_step(f"Error: {e}", "failed")
        return False

def test_email_send_simulation():
    """Test email send (simulation without actual sending)"""
    test_step("Testing email.send method (simulation)", "running")
    
    try:
        from src.mcp_client import MCPClient
        
        client = MCPClient("email", transport="stdio")
        
        # Call tools.list to verify server works
        result = client.call("tools.list", {})
        
        if isinstance(result, dict):
            test_step("Email methods available via MCP", "passed")
            
            # Show available methods
            tools = result.get("tools", [])
            methods = [t.get("name", "") for t in tools]
            print(f"   Available: {', '.join(methods[:5])}...")
            return True
        else:
            test_step("Unexpected result format", "failed")
            return False
            
    except FileNotFoundError:
        test_step("MCP server not found (expected if not configured)", "pending")
        print(f"   {Colors.WARNING}Note: Run 'pip install google-api-python-client google-auth-oauthlib' for full functionality{Colors.ENDC}")
        return True  # Not a failure, just needs setup
    except Exception as e:
        test_step(f"Error: {e}", "failed")
        return False

def test_task_processor_integration():
    """Test TaskProcessor has MCP integration"""
    test_step("Testing TaskProcessor MCP integration", "running")
    
    try:
        from src.claude_skills.ai_employee_skills.processor import TaskProcessor
        
        # Check if _send_email_response method exists
        processor = TaskProcessor(vault_path="obsidian_vault")
        
        if hasattr(processor, '_send_email_response'):
            test_step("TaskProcessor has MCP integration", "passed")
            return True
        else:
            test_step("TaskProcessor missing MCP method", "failed")
            return False
            
    except Exception as e:
        test_step(f"Error: {e}", "failed")
        return False

def test_json_rpc_format():
    """Test JSON-RPC 2.0 format"""
    test_step("Testing JSON-RPC 2.0 format", "running")
    
    try:
        # Valid JSON-RPC 2.0 request
        request = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "email.send",
            "params": {
                "to": "test@example.com",
                "subject": "Test",
                "body": "Hello"
            }
        }
        
        # Verify format
        assert request["jsonrpc"] == "2.0"
        assert "id" in request
        assert "method" in request
        assert "params" in request
        
        test_step("JSON-RPC 2.0 format correct", "passed")
        return True
        
    except Exception as e:
        test_step(f"Format error: {e}", "failed")
        return False

def main():
    """Run all MCP tests"""
    print_banner()
    
    print(f"{Colors.BOLD}Test Suite: Universal MCP Implementation{Colors.ENDC}\n")
    
    results = {
        "passed": 0,
        "failed": 0,
        "pending": 0
    }
    
    # Run tests
    tests = [
        test_json_rpc_format,
        test_mcp_client_import,
        test_mcp_client_instantiation,
        test_mcp_server_direct,
        test_email_send_simulation,
        test_task_processor_integration
    ]
    
    for test in tests:
        try:
            result = test()
            if result:
                results["passed"] += 1
            else:
                results["failed"] += 1
        except Exception as e:
            print(f"   {Colors.FAIL}Unexpected error: {e}{Colors.ENDC}")
            results["failed"] += 1
        print()
    
    # Summary
    print(f"\n{Colors.BOLD}{'='*60}{Colors.ENDC}")
    print(f"{Colors.BOLD}Test Summary{Colors.ENDC}")
    print(f"{Colors.BOLD}{'='*60}{Colors.ENDC}")
    print(f"  Passed:  {Colors.OKGREEN}{results['passed']}{Colors.ENDC}")
    print(f"  Failed:  {Colors.FAIL}{results['failed']}{Colors.ENDC}")
    print(f"  Pending: {Colors.WARNING}{results['pending']}{Colors.ENDC}")
    print()
    
    if results["failed"] == 0:
        print(f"{Colors.OKGREEN}✅ All tests passed! Universal MCP is operational.{Colors.ENDC}")
        print()
        print(f"{Colors.BOLD}Next Steps:{Colors.ENDC}")
        print("  1. Configure Gmail credentials for full email functionality")
        print("  2. Add MCP servers to Claude/Qwen/Gemini/Codex config")
        print("  3. Test with actual AI agents")
        print()
        return True
    else:
        print(f"{Colors.FAIL}❌ Some tests failed. Check errors above.{Colors.ENDC}")
        print()
        return False

if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)

File: src/mcp-servers/test-mcp.js
#!/usr/bin/env node
/**
 * Test MCP Servers
 * Tests Odoo and Email MCP servers
 */

import { spawn } from 'child_process';
import { setTimeout } from 'timers/promises';

console.log('='.repeat(70));
console.log('ELYX MCP SERVER TEST');
console.log('='.repeat(70));
console.log();

/**
 * Test Odoo MCP Server
 */
async function testOdooMCP() {
  console.log('[TEST 1] Odoo MCP Server');
  console.log('-'.repeat(70));
  
  return new Promise((resolve) => {
    const odoo = spawn('node', ['src/mcp-servers/odoo-mcp/index.js'], {
      env: {
        ...process.env,
        ODOO_URL: process.env.ODOO_URL || '',
        ODOO_DB: process.env.ODOO_DB || '',
        ODOO_USERNAME: process.env.ODOO_USERNAME || '',
        ODOO_PASSWORD: process.env.ODOO_PASSWORD || '',
        ODOO_API_KEY: process.env.ODOO_API_KEY || ''
      }
    });
    
    let output = '';
    let error = '';
    
    odoo.stderr.on('data', (data) => {
      error += data.toString();
      console.log('[ODOO]', data.toString().trim());
    });
    
    odoo.stdout.on('data', (data) => {
      output += data.toString();
    });
    
    // Give it 3 seconds to initialize
    setTimeout(3000).then(() => {
      odoo.kill();
      
      if (error.includes('Authenticated as user')) {
        console.log('[OK] Odoo MCP: Authentication successful');
        resolve(true);
      } else if (error.includes('Error') || error.includes('Failed')) {
        console.log('[ERROR] Odoo MCP: Authentication failed');
        resolve(false);
      } else {
        console.log('[OK] Odoo MCP: Started successfully');
        resolve(true);
      }
    });
  });
}

/**
 * Test Email MCP Server
 */
async function testEmailMCP() {
  console.log();
  console.log('[TEST 2] Email MCP Server');
  console.log('-'.repeat(70));
  
  return new Promise((resolve) => {
    const email = spawn('node', ['src/mcp-servers/email-mcp/index.js'], {
      env: process.env
    });
    
    let output = '';
    let error = '';
    
    email.stderr.on('data', (data) => {
      error += data.toString();
      console.log('[EMAIL]', data.toString().trim());
    });
    
    email.stdout.on('data', (data) => {
      output += data.toString();
    });
    
    // Give it 3 seconds to initialize
    setTimeout(3000).then(() => {
      email.kill();
      
      if (error.includes('Gmail initialized successfully')) {
        console.log('[OK] Email MCP: Gmail initialized');
        resolve(true);
      } else if (error.includes('Error') || error.includes('Failed')) {
        console.log('[WARNING] Email MCP: Gmail not initialized (check credentials)');
        resolve(false);
      } else {
        console.log('[OK] Email MCP: Started successfully');
        resolve(true);
      }
    });
  });
}

/**
 * Main test runner
 */
async function main() {
  console.log('Starting MCP server tests...\n');
  
  // Test Odoo MCP
  const odooOk = await testOdooMCP();
  
  // Test Email MCP
  const emailOk = await testEmailMCP();
  
  console.log();
  console.log('='.repeat(70));
  console.log('TEST RESULTS');
  console.log('='.repeat(70));
  console.log(`Odoo MCP:    ${odooOk ? '[OK]' : '[FAIL]'}`);
  console.log(`Email MCP:   ${emailOk ? '[OK]' : '[FAIL]'}`);
  console.log('='.repeat(70));
  
  if (odooOk && emailOk) {
    console.log();
    console.log('[SUCCESS] All MCP servers are working!');
    console.log();
    console.log('Next steps:');
    console.log('1. Add MCP servers to ~/.config/claude-code/mcp.json');
    console.log('2. Restart Claude Code');
    console.log('3. Test MCP tools in Claude Code');
  } else {
    console.log();
    console.log('[WARNING] Some MCP servers need attention');
    console.log();
    console.log('Troubleshooting:');
    console.log('- Odoo: Check credentials in .env');
    console.log('- Email: Ensure gmail_credentials.json and gmail_token.json exist');
  }
  
  process.exit(odooOk && emailOk ? 0 : 1);
}

main().catch(console.error);
